### PR TITLE
Added a semantic check for the witness problem

### PR DESCRIPTION
**Pull request by azreika**
The semantic check checks for arguments that are grounded by an aggregator's inner scope, but appear ungrounded in the outer scope.

Fixes issues #128 and #421.

E.g. suppose we have the program:

```
.decl a(x:number, y:symbol)
a(1,"x").
a(10,"y").
a(100, "z").

.decl b(x:number, y:symbol)
b(x,y) :- x = max z : { a(z,y) }.

.output b()
```

Semantically, we're trying to find the tuple `a(x,y)` where `x` attains its maximum value. The problem here, though, is that Soufflé currently doesn't actually store aggregator witnesses. For example, in the clause `b(x,y) :- x = max z : { a(z,y) }.` above, the witness environment consists of the variables `z` and `y`. The variable `y` is grounded only in the inner scope of the max aggregator, and so can't be used outside ungrounded since Soufflé isn't storing the witness information.

There was no semantic check for this previously - the problem either goes by unnoticed with no error or just crashes. For instance, the expected output of the above program is the pair `(100, "z")` if witnesses were permitted, but when run locally the program outputted the pair `(100, "x")`. With the semantic check in this pull request, the program will output the following error message instead:

```
Error: Witness problem: argument grounded by an aggregator's inner scope is used ungrounded in outer scope in file witness.dl at line 7
b(x,y) :- x = max z : { a(z,y) }.
----^-----------------------------
1 errors generated, evaluation aborted
```

In the future, can possibly extend to copy the witness environment to the outer scope, though this will probably affect performance. Using the witness environment in the outer scope never makes sense for the `count` or `sum` aggregators though.

### CODE REVIEW

**Comment by mmcgr** [comment]
A test for this in tests/semantic, such as your example, would be helpful. That would avoiding breaking this feature in future.

There was discussion of making this work for min/max. If it's decided to not do that, then remove the commented out test mentioned in #128. If there's no decision, or we do plan to support it, use `count` in the example, rather than `max` so that the test remains valid.

I think this should also be marked as fixes #421

**Comment by b-scholz** [comment]
Good point. We have plenty of negative examples from the raised issues. Perhaps we add them. I believe there is also an example in the slides of the tutorial.

File: tests/semantic/witness_check2/witness_check2.dl [hunk 1]
```diff
@@ -0,0 +1,32 @@
+.decl b(x:number)
```
Comment (mmcgr) [inline tests/semantic/witness_check2/witness_check2.dl]: Our usual copyright header should be at the start. After that, a brief comment about the point of the test would be good.

File: tests/semantic/witness_check/witness_check.dl [hunk 2]
```diff
@@ -0,0 +1,9 @@
+.decl A(x:number, y:symbol)
```
Comment (mmcgr) [inline tests/semantic/witness_check/witness_check.dl]: Does this test add anything not tested for in witness_check2?
If no, then remove it. If yes, add the header, as for witness_check2.

**Comment by mmcgr** [comment]
After discussion with Bernhard, min/max should stay as invalid. In which case this fixes #128 and the tests tests/evaluation/count_sccs2/count_sccs2.dl can be removed.

**Pull request merged**